charms/istio-pilot/src/charm.py

```diff
@@ -115,36 +163,48 @@ def get_kwargs(version, route):
             if 'namespace' not in kwargs:
                 kwargs['namespace'] = self.model.name
 
+            prefix = kwargs["prefix"]
+            kwargs.setdefault("rewrite", prefix)
+            if prefix == "/":
+                kwargs["unit_prefix"] = "/unit-{}/"
+            elif prefix.endswith("/"):
+                kwargs["unit_prefix"] = prefix[:-1] + "-unit-{}/"
+            else:
+                kwargs["unit_prefix"] = prefix + "-unit-{}"
+
+            kwargs["unit_nums"] = sorted(unit.name.split("/")[-1] for unit in rel.units)
+
             return kwargs
 
         virtual_services = ''.join(
-            t.render(**get_kwargs(ingress.versions[app.name], route))
-            for ((_, app), route) in routes
+            t.render(**get_kwargs(rel, ingress.versions[app.name], route))
+            for ((rel, app), route) in routes.items()
         )
 
-        t = self.env.get_template('gateway.yaml.j2')
-        gateways = self.model.config['default-gateways'].split(',')
-        gateways = ''.join(t.render(name=g) for g in gateways)
-
-        manifests = [virtual_services, gateways]
-        manifests = '\n'.join([m for m in manifests if m])
-
-        subprocess.run(
-            [
-                './kubectl',
-                'delete',
-                'virtualservices,gateways',
-                f'-lapp.juju.is/created-by={self.model.app.name}',
-                '-n',
-                self.model.name,
-            ],
-            check=True,
-        )
-        subprocess.run(
-            ["./kubectl", "apply", "-f-"],
-            input=manifests.encode('utf-8'),
-            check=True,
+        self._kubectl(
+            'delete',
+            'virtualservices,destinationrules',
+            f'-lapp.juju.is/created-by={self.app.name}',
         )
+        if routes:
+            self._kubectl("apply", "-f-", input=virtual_services)
+
+        # Send URL(s) back
```

Comment:
What happens if a cluster exposes multiple ingresses for the same set of services, e.g. `kubeflow.company.com` and `kubeflow.company.vpn`/`kubeflow-vpn.company.com`? Or a situation like `old.reddit.com`/`www.reddit.com`/`my-subreddit.reddit.com`, where the prefix has meaning, but is handled by the same set of servers.

It really doesn't seem like `ingress` should be handling any notion of public URLs. As it is currently, an `ingress` requirer says to the provider "I'd like to be routed to, here's how to talk to me", and doesn't care which public URL is used, as that's up to how the provider is configured.

Introducing the notion of public URL into `ingress` forces charmed services to know/care about things that they really shouldn't.

Reply:
Agreed.  The relation is just "tell me what my ingress address is" and whether that's public with an external hostname or just across internal clusters, that's up to the ingress provider (Istio, in this case) to manage.  Typically, that would be done by config on either whatever is providing the `LoadBalancer` services their external address, or config on the provider charm to override it with something that's managed entirely externally (like a public DNS entry).